---
title: "R Notebook"
output: html_notebook
---


data got from this papaer: https://pubmed.ncbi.nlm.nih.gov/25730472/
data can be accessed: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE60450

```{r, warning=FALSE, eval=FALSE}
# Installing packages if not already installed
if (!requireNamespace("BiocManager"))
    install.packages("BiocManager")
list.of.packages <- c("RColorBrewer", "NMF", "ggplot2", "EnhancedVolcano", "DESeq2")
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)
```
These are the packages this tutorial will use.
```{r, eval=FALSE,warning=FALSE}
library(RColorBrewer)
library(NMF)
library(ggplot2)
library(EnhancedVolcano)
library(DESeq2)
```
Setting the working directory
```{r}
setwd('C:/Users/Mischa/Documents/Uni Masters/Module 6 - Group proj/Blogpost')
```

## Reading in the data

First to read in the data, ensure that your working directory is set to the file that the data is in. 
```{r}
# Read the data into R
seqdata <- read.delim("GSE60450_Lactation-GenewiseCounts.txt", stringsAsFactors = FALSE)
# Read the sample information into R
sampleinfo <- read.delim("SampleInfo_Corrected.txt", stringsAsFactors = TRUE)
```

Now we can have a look at the data we have imported
```{r}
head(seqdata)
head(sampleinfo, 10)
```
Seqdata has information about the genes, the first two columns contain geneID and length.
Sampleinfo has basic information about each sample that is required for analysis.


## Formatting 

For analysis we want a dataframe with geneID as rownames and column names matching sampleinfo$SampleName, containing only the gene counts.
```{r}
seq_counts <- seqdata[,-c(1,2)]
rownames(seq_counts) <- seqdata[,1]
# Taking only the first 7 characters of column name
colnames(seq_counts) <- sampleinfo$SampleName
```
Now we can look at the formatted data.
```{r}
head(seq_counts)
```

# Converting to DGEList object

DGEList is an object used in RNA analysis to store gene count data.
```{r}
DGEcount <- DGEList(seq_counts)
# Look at the data
head(DGEcount)
```
You can see this is a list of length 2, storing counts and samples in each element. Group and library size are stored in samples while count data is stored in counts. We can change the group column to store the cell type and status of each sample. 
```{r}
group <- factor(paste(sampleinfo$CellType, sampleinfo$Status, sep='_'))
DGEcount$samples$group <- group
head(DGEcount)
```

## Data filtering

When conducting RNA analysis the researcher must remove genes that show consistently low level expression, this is because they reduce differential expression power and have little statistical value. For this the edgeR package is used. 

This data has two samples in each group, and so the low expression threshold will be below 0.5 counts per million in at least two samples. A general CPM rule is to use one that represents a count of 10. This can be filtered with the filterByExpr function in edgeR.
```{r}
keep <- filterByExpr(DGEcount, group = group, min.total.count=10)
DGEcount <- DGEcount[keep, keep.lib.sizes=FALSE]
```

## Quality control

It is important to check that there is no large differences between library size (samples).
```{r}
ggplot(DGEcount$samples, aes(x = colnames(DGEcount), y=lib.size)) + geom_bar(stat = 'identity', fill = '#00abff', color = 'black') + 
  theme(axis.text.x = element_text(angle = 90)) +
  xlab('Sample ID') +
  ylab('Library size (millions)') +
  ggtitle('Library size')
```
Next we should see whether the counts are distributed appropriately. For this we use cpm function to plot on a log2 scale. 

```{r}
# Obtain log counts
log_count <- cpm(DGEcount, log=TRUE)
# Save median value for the abline
abline <- median(log_count)
# Plotting the counts with an abline and data labels.
Unnormalised_boxplot <- ggplot(stack(as.data.frame(log_count)), aes(x = ind, y = values)) +
  geom_boxplot(fill = '#00abff') +
  theme(axis.text.x = element_text(angle = 90)) + 
  geom_hline(yintercept = abline, color = 'red') +
  xlab('Sample ID') +
  ylab('Counts per million (log2)') +
  ggtitle('LogCMP values') 
Unnormalised_boxplot
```

## Data Visualisation

## Multidimensional Scaling Plot

MDS plot is one of the most important for RNA-Seq analysis, these plots visualise PCA without data reduction, it is a purely mapping plot. In this section we will compare a simple MDS plot to a more visually informative one by adding some aesthetic code. 
```{r}
plotMDS(DGEcount)
```
This plot shows us that there are in fact pairings within the data, but does not show whether cell type or status has any impact, as all samples are overlapped and not coloured. This plot can become more informative with aesthetic changes. 
```{r}
# Allowing side-by-side plots of Cell type and Status
par(mfrow=c(1,2))
col_celltype <- c("#9932CC", "#FDD20EFF")[sampleinfo$CellType]
col_status <- c("#00AFBB", "#E7B800", "#FC4E07")[sampleinfo$Status]
# Plotting cell type MDS graph
plotMDS(DGEcount, col=col_celltype)
legend("topleft",fill=c("#9932CC","#FDD20EFF"),legend=levels(sampleinfo$CellType))
title("Cell type")
# Plotting sample Status MDS graph
plotMDS(DGEcount, col=col_status)
legend("topleft",fill=c("#00AFBB","#E7B800","#FC4E07"),legend=levels(sampleinfo$Status),cex=0.8)
title("Status")

```
From this you can see that samples are grouped in pairs, these pairings are replicates from the same group. The further away points are from each other the more difference there is in the expression. Through this visualisation we can see that the Cell type and Status has an impact on RNA expression counts. 
These two plots can be further simplified by unifying them into a single graph, as shown below.
```{r}
colors <- c("#00AFBB", "#E7B800", "#FC4E07")
colors <- colors[as.numeric(sampleinfo$Status)]

shapes = c(16, 17) 
shapes <- shapes[as.numeric(sampleinfo$CellType)]
par(mfrow=c(1,1))
plotMDS(DGEcount,col=colors,cex=2, pch =shapes)
legend("topright",legend=levels(sampleinfo$Status), col = c("#00AFBB", "#E7B800", "#FC4E07") ,pch=16)
legend("bottom",legend=levels(sampleinfo$CellType),pch=c(16,17))
```
The combined graph shows the data much more clearly, as the difference between cell type and status is clearly displayed on each graph. Aesthetic changes have made a complicated subject easily understandable for non-data scientists. 
Additionally, you could chose to do a PCA plot, which is very similar to a basic MDS plot but requires a few more lines of code:
```{r}
# # transforming it so that samples are on the row and chemicals are columns
pca_x <- t(seq_counts)
pca_x <- pca_x[ , which(apply(pca_x, 2, var) != 0)]
# # doing the pca
pca_res1 <- prcomp(pca_x, center = TRUE, scale = TRUE)
# making the dataframe
pc_df1 <- data.frame(pca_res1$x, Celltype = sampleinfo$CellType, Status = sampleinfo$Status)
# plotting it
library(RColorBrewer)
mycolours <- c('#00A4CCFF', '#F95700FF')
ggplot(pc_df1, aes(PC1, PC2, shape = Status)) + 
  geom_point(aes(fill = Celltype), size = 6) + ggtitle('PCA of Cell type and Status using Mouse data') + scale_shape_manual(values=c(21,24, 23)) + guides(fill = guide_legend(override.aes = list(shape = 21)))
```


## Heatmap Hierarchical clustering

Heatmaps are commonly used in Differential expression analysis to convey relationships between genes. For this we will plot the most variable genes to not overwhelm the plot.

```{r}
# Estimatinve variance per row in log_count matrix
variable_genes <- apply(log_count, 1, var)
# Take gene names of the 500 most variable genes
selected_var <- names(sort(variable_genes, decreasing=TRUE))[1:500]
# Subset logcounts matrix to take only the top 500 genes
highly_variable_genes <- log_count[selected_var,]
# Plot the data
heatmap(highly_variable_genes)
```

This heatmap is okay, but it can be improved with aesthetics. Right now we can see that a subset of genes are differentially expressed in the different samples, but it is quite cluttered. 
```{r}
## Get some nicer colours
mypalette <- brewer.pal(11,"RdYlBu")
morecols <- colorRampPalette(mypalette)
col.cell <- c("purple","orange")[sampleinfo$CellType]
# Reducing the number of genes in the graph
highly_variable_genes_100 <- log_count[names(sort(variable_genes, decreasing=TRUE))[1:100],]
# plotting
heatmap(highly_variable_genes_100, col=rev(morecols(50)), main="Top 100 most variable genes across samples",ColSideColors=col.cell,scale="row")
```
This plot is much better, but more annotation can be added to improve it further

```{r}
# Specify colors
ann_colors = list(CellType = c('purple', 'orange'), Status = c('#28334AFF', '#FBDE44FF', '#F65058FF'))
mypalette <- brewer.pal(11,"RdYlBu")
morecols <- colorRampPalette(mypalette)
# Plot
par(mfrow=c(1,1))
aheatmap(highly_variable_genes_100,col=rev(morecols(50)),main="Top 500 most variable genes across samples",annCol=sampleinfo[, 3:4],annColors=ann_colors,labCol=group, scale="row")

```

This plot is far more informative than the previous iterations. With colours explaining the cell type and status along with clear sample labels. We can see a clear distinction between cell type, but less of a distinction between cell status.

## Differential Expression Analysis

First we will create a design matrix. For this we will test the differences in cell type and status separately.

```{r}
# First we must normalise for composition bias
DGEcount <- calcNormFactors(DGEcount)
# Make a design matrix with no intercept term
designmat <- model.matrix(~ 0 + group)
colnames(designmat) <- levels(group)
```

Once the design matrix has been made we will Voom transform the data, this adjusts library sizes automatically. We can test the output through a boxplot to better explain the normalisation.
```{r}
# Normalisation
transformed <- voom(DGEcount, designmat, plot = TRUE)
names(transformed)
```
```{r}

par(mfrow=c(1,2))

# Test for normalisation with a boxplot
ggplot(stack(as.data.frame(transformed)), aes(x = ind, y = values)) +
  geom_boxplot(fill = '#00abff') +
  theme(axis.text.x = element_text(angle = 90)) + 
  geom_hline(yintercept = abline, color = 'red') +
  xlab('Sample ID') +
  ylab('Counts per million (log2)') +
  ggtitle('LogCMP values') 
# Plot the earlier boxplot to compare
Unnormalised_boxplot
```

Now we can test for differential expression on our normalised data using a linear model. lmFit approximates group means using the design matrix and gene-wise variance. makeContrasts function lets you specify the differential expression you are interested in, for this blog we will look at lactating vs pregnant groups. 

```{r}
# Fitting the linear model
lm <- lmFit(transformed)
names(lm)
# Applying make Contrasts
cont_df <- makeContrasts(Basal_lacVsPred=basal_pregnant - basal_lactate,levels=designmat)
# Viewing the contrast dataframe
cont_df
```
The contrast dataframe tells the limma package which columns of the design matrix to use in the comparison test.
```{r}
# Apply the contrast dataframe to the linear model to get the comparison
fit_contrast <- contrasts.fit(lm, cont_df)
dim(fit_contrast)
# Use empirical Bayes shrinkage of variance to better plot
fit_contrast <- eBayes(fit_contrast)
```
You can get a differential expression summary using decideTests in limma
```{r}
fit_summary <- decideTests(fit_contrast)
summary(fit_summary)
```

## Plotting differential expression

Volcano plots are commonly used in differential expression analysis as a visual way to tell how may genes are differentially expressed. 
```{r}
plotMD(fit_contrast, coef = 1, status = fit_summary, values = c(-1, 1), hl.col=c("blue","red"))

```

The volcano plot shows 

```{r}
volcanoplot(fit_contrast,coef=1,highlight=100,names=fit_contrast$genes$SYMBOL, main="Basal_lacVsPred")
```

## Enhanced volcano plots using DESEQ analysis
```{r}
dge <- DGEList(round(seq_counts), group = sampleinfo$Status)
dds <- as.DESeqDataSet(dge)
sdd <- DESeq(dds)
res <- results(sdd)
resultsNames(sdd)
resLFC2 <- lfcShrink(sdd, coef = 'group_pregnant_vs_lactate', type='apeglm')



EnhancedVolcano(resLFC2,
    lab = rownames(resLFC2),
    x = 'log2FoldChange',
    y = 'pvalue')

```

